### PR TITLE
Possible fix for SbtConnectionProxySpec

### PR DESCRIPTION
**Pull request by corruptmemory**
I think this should have better inter-thread sharing sematics.

### CODE REVIEW

**Comment by corruptmemory** [comment]
The real test is if Travis agrees with me.

File: actor-client/src/test/scala/sbt/client/actors/FakeSbtClient.scala [hunk 1]
```diff
@@ -147,7 +147,7 @@ final class FakeSbtClient(refFactory: ActorRefFactory,
   requestExecutionKey: (protocol.ScopedKey, Option[(Interaction, ExecutionContext)]) => Future[Long] = FakeSbtClient.emptyRequestExecutionKey,
   _cancelExecution: Long => Future[Boolean] = FakeSbtClient.emptyCancelExecution,
   val uuid: java.util.UUID = UUID.randomUUID()) extends SbtClient {
-  private final val closed = new AtomicBoolean(false)
+  @volatile private var closed: Boolean = false
```
Comment (jsuereth) [inline actor-client/src/test/scala/sbt/client/actors/FakeSbtClient.scala]: why is volatile better?

File: actor-client/src/test/scala/sbt/client/actors/FakeSbtClient.scala [hunk 2]
```diff
@@ -239,13 +239,18 @@ final class FakeSbtClient(refFactory: ActorRefFactory,
     ??? // TODO stubbed until someone actually wants to use it
 
   def requestSelfDestruct(): Unit = ()
-  def isClosed: Boolean = closed.get()
+  def isClosed: Boolean = this.synchronized {
+    closed
+  }
 
   def close(): Unit = {
+    this.synchronized {
```
Comment (jsuereth) [inline actor-client/src/test/scala/sbt/client/actors/FakeSbtClient.scala]: NOt sure how this changes things...

File: actor-client/src/test/scala/sbt/client/actors/SbtConnectionProxySpec.scala [hunk 3]
```diff
@@ -34,7 +34,7 @@ object SbtConnectionProxySpec {
       } finally {
         cp ! Close(testActor)
         Assert.assertEquals(expectMsgType[Closed.type], Closed)
-        if (testClosed) Assert.assertEquals(client.isClosed, true)
+        if (testClosed) Assert.assertTrue("Ensure the client is closed", client.isClosed)
```
Comment (jsuereth) [inline actor-client/src/test/scala/sbt/client/actors/SbtConnectionProxySpec.scala]: I think I'd just add a delay here, or a timeout...

File: actor-client/src/test/scala/sbt/client/actors/FakeSbtClient.scala (hunk 1 above)
Comment (corruptmemory) [inline actor-client/src/test/scala/sbt/client/actors/FakeSbtClient.scala]: No idea. Just experimenting. Agreed there should be no difference.

File: actor-client/src/test/scala/sbt/client/actors/FakeSbtClient.scala (hunk 2 above)
Comment (corruptmemory) [inline actor-client/src/test/scala/sbt/client/actors/FakeSbtClient.scala]: Neither am I. Clearly there's something about OpenJDK 6 that is funky.

File: actor-client/src/test/scala/sbt/client/actors/SbtConnectionProxySpec.scala (hunk 3 above)
Comment (corruptmemory) [inline actor-client/src/test/scala/sbt/client/actors/SbtConnectionProxySpec.scala]: That's exactly what I was going to try this morning. :-)

**Comment by havocp** [comment]
maybe add some logging to each step in the sequence of events that's supposed to happen, and see if we can understand what goes wrong?

**Comment by corruptmemory** [comment]
That's such a crazy idea it might just work! :-)

Yeah, going to get around to that as well.  One concern is that this may be some kind of subtle timing error that the extra logging may mask.

The test is currently failing for openjdk 6; I'm going to install that locally to see if I can make it happen on my machine.

**Comment by corruptmemory** [comment]
Well, found the issue.  Something about the test is broken. Nothing related to JDK versions or anything.  Will fix sometime today.

**Comment by corruptmemory** [comment]
Nope, not just a wrong test.  Wrong code.  HA! Classic problem in code evolution: there is a small amount of code left over from an earlier version of the SbtConnectionProxy that is now wrong.  Yay tests! :-)

**Pull request merged**